src/util.ts

```diff
@@ -89,6 +89,30 @@ export function createOutputRetainingCompilerHost(
   }
 }
 
+/**
+ * Returns the declarations for the given decorator.
+ */
+export function getDecoratorDeclarations(decorator: ts.Decorator, typeChecker: ts.TypeChecker) {
+  // Walk down the expression to find the identifier of the decorator function.
+  let node: ts.Node = decorator;
+  while (node.kind !== ts.SyntaxKind.Identifier) {
+    switch (node.kind) {
+      case ts.SyntaxKind.Decorator:
+      case ts.SyntaxKind.CallExpression:
+        node = (node as ts.Decorator | ts.CallExpression).expression;
+        break;
+      default:
+        throw new Error();
```

Comment:
This does refer to foo.bar. The api worked as follows, if `shouldDownlevel` returns true, we call `process` on it.

When I last touched this file, I tried to add PAE, but it didn't work because `process` failed to do anything. I found it surprising too and added the comment. Technically we should be able to handle simple propaccess.

I am against throwing an error in decorators we cannot statically analyze. Remember, the original TS team design is to have dynamic decorators (like python). We are doing some static shenanigans to have annotations, but the semantics of annotations are not defined by the language and we keep on iterating on them (your cl adding even more custom semantics). The annotation rewrite should quietly continue if it cannot prove that the decorator comes from an annotation.

```javascript
function getFoo() {
 return isPrime(2^100 + 1) ? anAnnotation : aDecorator;
}

@getFoo()
class X { ... }
```